### PR TITLE
Code cleanup

### PR DESCRIPTION
**Pull request by vcolin7**
- Reviewed all JavaDoc comments and corrected/reformatted where required.
- Moved and changed artifact's names where appropriate for the sake of clarity and consistency.
- Removed unnecessary classes.
- Changed the visibility of some code artifacts that should not be exposed to customers.
- Consolidated some interfaces and their implementations as described in [this commit](https://github.com/Azure/azure-sdk-for-android/commit/bf8ed6437d7bf5e6ef56154967be0edafbf89aea) for the sake of having a clearer and cleaner public API.

### CODE REVIEW

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/Base64Util.java [hunk 1]
```diff
@@ -9,78 +9,76 @@
 /**
  * Utility for Base64 encoding and decoding.
  */
-public final class Base64Util {
+public interface Base64Util {
```
Comment (anuchandy) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/util/Base64Util.java]: We shouldn't turn this to an interface given there is no use-case to implement it.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/logging/AndroidClientLogger.java [hunk 2]
```diff
@@ -61,48 +63,80 @@ public void setLogLevel(@LogLevel int logLevel) {
...
+    @Override
     public void warning(String message) {
         if (LOG_LEVEL_WARNING>= this.logLevel) {
             Log.w(tag, message);
         }
     }
 
+    /**
+     * {@inheritDoc}
+     */
```
Comment (anuchandy) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/util/logging/AndroidClientLogger.java]: nit; given this is internal type we can remove inherit-javadoc for all overriden methods

**Comment by JonathanGiles** [comment]
Did this PR get the go-ahead to merge? I don't see any approvals which always makes me nervous.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/interceptor/DeviceInformation.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+/**
+ * Interface for an object that contains application, platform and locale information.
+ */
+public interface DeviceInformation {
+    /**
+     * Creates and returns a default {@link DeviceInformation} in the form of an {@link AndroidInformation} instance.
+     *
+     * @param configuration The application's context.
+     * @param context       The system configuration.
+     * @return An {@link AndroidInformation} instance.
```
[the remark this answers is not in the document]
Comment (vcolin7) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/http/interceptor/DeviceInformation.java]: Removed this class and went back to using separate providers. Will add a way of getting a default provider while making the implementation package private in a separate PR.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/interceptor/AndroidInformation.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    @Override
+    public String getModel() {
+        return Build.MODEL;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int getOsVersion() {
```
[the remark this answers is not in the document]
Comment (vcolin7) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/http/interceptor/AndroidInformation.java]: Good to know, thanks!

**Comment by vcolin7** [comment]
@JonathanGiles This was blocking some unit testing work I am busy with, so I talked about merging it with @bsiegel but forgot to ask him for approval here though. I was also careful of solving some merging conflicts in the process.

**Comment by JonathanGiles** [comment]
The fact that a code cleanup PR can block unit testing work is why I have issue with this PR in the first instance - a code cleanup PR should not be introducing new API. That these are being conflated but presented as code cleanup shouldn't be happening. They should be separate pull requests. 

I also worry about changing formatting without first reaching consensus on what code style should be used, e.g.
 * There appears to be mismatched JavaDoc tags - either JavaDoc was not run before this was merged or we don't have our JavaDoc settings strict enough to enforce valid HTML. 
* Indentation of the parameter text in the javadoc so that they are aligned - I would rather we didn't do this.
* Reflowing arguments into a method such that there is one argument per line. Again, I would rather not follow this style.

**Comment by vcolin7** [comment]
@JonathanGiles Although I agree that we should separate concerns when it comes to PRs in and be very clear about the purpose of each one, I think we are at a stage where we can be flexible with some of our rules in favor of efficiency to have a working preview that can also be presented to the architecture board in a timely fashion, which is the purpose behind trying to unify the coding style. Whether the choices I made in style are something are what would be preferred or not could be discussed separately, I think.

**Comment by JonathanGiles** [comment]
I don't agree - there is no value in making changes to style in advance of agreeing on the style, and the code style has no impact on the architectural board reviews (because we don't review code style). It is just work that may need to be reverted later on.

**Comment by vcolin7** [comment]
Fair point. In hindsight, I think the purpose behind doing these changes was to get ahead of the curve and propose something that looked decent and consistent enough so new changes wouldn't have to be drastic. I didn't stop and consider that it makes it hard for everyone that is reviewing to keep different issues together in one big PR. Furthermore, this conversation made me realize that there is a lot I have to learn about working in an open source project and that I don't yet fully understand all the factors that are considered in the architectural review meetings.

If you don't mind, I would like us to have a chat on this matters at some point later in the week.

Thanks for the feedback :)

File: sdk/core/azure-core/src/main/java/com/azure/android/core/internal/http/rest/RetrofitAPIClient.java [hunk 5]
```diff
@@ -1,10 +1,10 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-package com.azure.android.core.implementation.http.rest;
+package com.azure.android.core.internal.http.rest;
 
-import com.azure.android.core.implementation.util.serializer.SerializerAdapter;
-import com.azure.android.core.implementation.util.serializer.SerializerEncoding;
+import com.azure.android.core.internal.util.serializer.SerializerAdapter;
+import com.azure.android.core.internal.util.serializer.SerializerEncoding;
```
[the remark this answers is not in the document]
Comment (vcolin7) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/internal/http/rest/RetrofitAPIClient.java]: Based on more recent discussions we'll stick with internal for now.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/Base64Util.java (hunk 1 above)
Comment (vcolin7) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/util/Base64Util.java]: This change was made to keep consistence with our CoreUtils implementation, which was decided to be an interface with only static methods as outlined in [this comment](https://github.com/Azure/azure-sdk-for-android/pull/128#discussion_r350471301).

**Comment by bsiegel** [comment]
Just wanted to chime in here, and first apologize as I was heads down and didn't get a chance to thoroughly respond earlier. I had taken a few quick passes at this PR before it got merged, but I didn't have a chance to focus in depth enough to write review comments and submit a GitHub review. This was primarily due to me focusing on other things and not realizing that this PR was blocking progress, but secondarily due to its overall size and variety of changes.

That said, Victor and I did chat offline about it and I was comfortable merging it in as-is. I don't have an opinion about what style we follow, as long as we are consistent and my understanding is that the style changes were purely to enforce consistency. I'm happy to punt the conversation about /what/ the style should be to a later time.

We did talk about making sure that in the future, PRs are smaller and more focused, with a single conceptual change. That will make it easier and faster to get the PR reviewed and merged, and avoid blocking important functional changes on bikeshedding topics like code formatting. We are definitely in a (very temporary) position where we can be a little loose with how intensely we review code going in, in support of being able to reach a first preview with the limited resourcing and time we have. However, that doesn't mean we should also be loose with /how/ we throw code over for review. In this example, being loose with what changes were included with this PR actually slowed us down and made us take longer to merge it, rather than improving efficiency.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/annotation/JsonFlatten.java [hunk 6]
```diff
@@ -3,21 +3,17 @@
...
- *
+ * Annotation used for flattening properties separated by '.', e.g., a property with JsonProperty value
+ * "properties.value" will have "value" property under the "properties" tree on the wire.
  */
 @Retention(RUNTIME)
 @Target({ElementType.ANNOTATION_TYPE, ElementType.TYPE})
 public @interface JsonFlatten {
-}
\ No newline at end of file
+}
```
[the remark this answers is not in the document]
Comment (vcolin7) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/annotation/JsonFlatten.java]: Will remove in a future PR, since this is already merged.

**Pull request merged**